Add an ADR for sync payload evolution (#5434)
mozilla/application-services: docs/adr/0008-sync-payload-evolution.md

# Handling the evolution of Sync payloads

* Status: Accepted
* Deciders: sync team, credentials management team
* Date: 2023-03-15

Technical Story:
* https://github.com/mozilla/application-services/pull/5434
* https://docs.google.com/document/d/1ToLOERA5HKzEzRVZNv6Ohv_2wZaujW69pVb1Kef2jNY

## Context and Problem Statement

Sync exists on all platforms (Desktop, Android, iOS), all channels (Nightly, Beta, Release, ESR) and is heavily used across all Firefox features.
Whenever there are feature changes or requests that potentially involve schema changes, there are not a lot of good options to ensure sync doesn’t break for any specific client.
Since sync data is synced from all channels, we need to make sure each client can handle the new data and that all channels can support the new schema.
Issues like [credit card failing on android and desktop release channels due to schema change on desktop Nightly](https://bugzilla.mozilla.org/show_bug.cgi?id=1812235)
is an example of such cases we can run into.
This document describes our decision on how we will support payload evolution over time.

> Note that even though this document existing in the application-services repoository, it should
> be considered to apply too all sync implementation, whether in this repository, in mozilla-central,
> or anywhere else.

Reviewer: ```suggestion
Note that even though this document exists in the application-services repository, it should
be considered to apply to all sync implementations, whether in this repository, in mozilla-central,
or anywhere else.
```

> or anywhere else.

Author: huh - so yeah, maybe I should actually just land this under https://firefox-source-docs.mozilla.org/services/sync/index.html, as I guess that's really where the canonical public sync docs should live - WDYT?

> or anywhere else.

Reviewer 2: That seems like a good place to me.

  or have support for new features in "new" versions, but which we still want to support without
  breakage and without the user perceiving data-loss. This is typically accepted to mean the
  current ESR version or later, but taking into account the slow update when new ESRs are released.

* An "old" version is any version before what we consider "recent".

## Decision Drivers

* It must be possible to change what data is carried by Sync to meet future product requirements.
* Both desktop and mobile platforms must be considered.
* We must not break "recent" Firefox installations, nor cause what the user might perceive as data-loss,
  when a "new" Firefox installation syncs, and vice-versa.
* Some degree of breakage for "old" Firefox installations when "new" or "recent" firefoxes sync
  might be considered acceptable if absolutely necessary.
* However, breakage of "new" or "recent" Firefoxes should "old" versions sync is *not* acceptable.
* Because such evolution should be rare, we do not want to set an up-front policy about locking out
  "old" versions just because they might have a problem in the future. That is, we want to avoid

>   a policy that dictates versions more than (say) 2 years old will break when syncing "just in case"

Author: I actually don't like "decision drivers" because I don't quite understand what it means :) If I renamed this to simply "requirements" (which I do :) then I think this para still belongs there, right? Or are you suggesting that it's misplaced because it's really just stating a non-requirement?

>   a policy that dictates versions more than (say) 2 years old will break when syncing "just in case"

Reviewer 2: The way I see it "Some degree of breakage for "old" Firefox installations when "new" or "recent" firefoxes sync might be considered acceptable if absolutely necessary." is the requirement, which is why we don't want to have a blanket policy of locking out older versions.  I would either remove the bullet or discuss locking out older versions as in the options section.  But I'm definitely open to keeping it here if you like it.

meeting the decision drivers.

## Pros and Cons of the Options

### A backwards compatible schema policy

A summary of this option is a policy by which:

* All sync engines should identify where in their payload new fields are likely to be added in the
  future, and when syncing these records, arrange to persist any fields from the payload which it
  does not understand. The next time that engine needs to upload that record to the storage server,
  it must arrange to add all such "unknown" fields back into the payload.

* Different engines must identify different locations where this might happen. For example, the
 `passwords` engine would identify the "root" of the payload, `addresses` and `creditcards` would
 identify the `entry` sub-object in the payload, while the history engine would probably identify
 *both* the root of the payload and the `visits` array.

* Fields can not change type, nor be removed for a significant amount of time. This might mean
  that "new" clients must support both new fields *and* fields which are considered deprecated
  by these "new" clients because they are still used by "recent" versions.

> The pros and conts:

Reviewer: ```suggestion
The pros and cons:
```

* Bad, because it's not possible to prove a proposed change meets the requirements - the policy
  is informal and requires good judgement as changes are proposed.

### A policy which prevents "recent" clients from syncing, or editing data

Proposals which fit into this category might have been implemented by (say) adding
a version number to the schema, and if clients did not fully understand the schema it would
either prevent syncing the record, or sync it but not allow editing it, or similar.

This was rejected because:

* The user would certainly perceive data-loss if we ignored the incoming data entirely.
* If we still wanted older versions to "partially" see the record (eg, but disallow editing) we'd
  still need most of the chosen option anyway - specifically, we could still never
  deprecate fields etc.
* The UI/UX of trying to explain to the user why they can't edit a record was deemed impossible
  to do in a satisfactory way.
* This would effectively penalize users who chose to use Nightly Firefoxes in any way. Simply
  allowing a Nightly to sync would effectively break Release/Mobile Firefox versions.

> ### A formal schema-driven process.

Reviewer 3: This is what protobuf does, but of course that's a land very far away from where we are :) Just reminded me of some reading I did

(https://earthly.dev/blog/backward-and-forward-compatibility/)

Ideally we could formally describe schemas, but we can't come up with anything here which
works with the constraints of supporting older clients - we simply can't update older released

> Firefoxes so they know how to work with the new schemas. We also couldn't some up with a solution

Reviewer: ```suggestion
Firefoxes so they know how to work with the new schemas. We also couldn't come up with a solution
```

where a schema is downloaded dynamically which also allowed the *semantics* (as opposed to simply
validity) of new fields to be described.

### Consider the sync payloads frozen and never change them.

A process where payloads are frozen was rejected because:

* The most naive approach here would not meet the needs of Firefox in the future.

* A complicated system where we started creating new payload and new collections
  (ie, freezing "old" schemas but then creating "new" schemas only understood by
  newer clients) could not be conceived in a way that still met the requirements,
  particularly around data-loss for older clients. For example, adding a credit-card
  on a Nightly version but having it be completely unavailable on a release firefox
  isn't acceptable.

> 

Reviewer 2: Maybe one more?

### Use separate collections for new data

We could store the new data in a separate collection.  For example define a `bookmarks2` collection where each record has the same `guid` as one in `bookmarks` alongside any new fields.  Newer clients use both collections to sync.

Pros and cons:

* Good, because it allows newer clients to sync new data without affecting recent or older clients
* Bad, because sync writes would lose atomicity without server changes.  We can currently write to a single collection in an atomic way, but don't have a way to write to multiple collections.
* Bad because this number of collections grows each time we want to add fields.
* Bad because it potentially leaks extra information to an attacker that gets access to the encrypted server records.  For example if we added a new collection for a single field, then the attacker could guess if that field was set or not based on the size of the encrypted record.
* Bad because it's difficult to handle nested data with this approach, for example adding a field to a history record visit.
* Bad because it has the same issue with dependent data as the chosen solution.

Author: That's what I was trying to get at in the bullet point "A complicated system where we started creating new payload and new collections", but yeah, yours goes into more detail, so I added it, thanks!

  asks coming down the line which require this capability.

## Considered Options

* A backwards compatible schema policy, consisting of  (a) having engines "round trip" data they
  do not know about and (b) never changing the semantics of existing data.
* A policy which prevents "recent" clients from syncing, or editing data, or other restrictions.
* A formal schema-driven process.
* Consider the sync payloads frozen and never change them.
* Use separate collections for new data

## Decision Outcome

Chosen option: A backwards compatible schema policy because it is very flexible and the only option
meeting the decision drivers.

## Pros and Cons of the Options

### A backwards compatible schema policy

A summary of this option is a policy by which:

* All sync engines should identify where in their payload new fields are likely to be added in the
  future, and when syncing these records, arrange to persist any fields from the payload which it
  does not understand. The next time that engine needs to upload that record to the storage server,
  it must arrange to add all such "unknown" fields back into the payload.

> * Different engines must identify different locations where this might happen. For example, the

Author: [this is replying to both comments above]

I'd like to see some room for best judgement here though - eg, as written, it sounds like we should also persist unknown fields in the top-level of an "address" payload, even though I believe the design is explicitly that new fields go in "entry" so we should never find them in the top-level? Ditto for tabs - as written, we should (a) persist them for tabs too, even though theory are only written by the owner, but must also persist them in the top of the payload, the `tabs` array, and `urlHistory` for each tab.

Another example in our future might be passkeys. IIUC, they are likely to be stored as a sub-object, and each site can have any number of passkeys associated with it. It's not clear to me now that we must support unknown fields in that sub-object, but if we make this section too prescriptive then we will be forcing it even if the use-case ends up not really existing. IOW, I think it's something that the implementers of passkeys need to *consider* when implementing the change, but not something this document should *insist* they do against their judgement at the time.

WDYT?

* If we still wanted older versions to "partially" see the record (eg, but disallow editing) we'd
  still need most of the chosen option anyway - specifically, we could still never
  deprecate fields etc.
* The UI/UX of trying to explain to the user why they can't edit a record was deemed impossible
  to do in a satisfactory way.
* This would effectively penalize users who chose to use Nightly Firefoxes in any way. Simply
  allowing a Nightly to sync would effectively break Release/Mobile Firefox versions.

### A formal schema-driven process.

Ideally we could formally describe schemas, but we can't come up with anything here which
works with the constraints of supporting older clients - we simply can't update older released
Firefoxes so they know how to work with the new schemas. We also couldn't come up with a solution
where a schema is downloaded dynamically which also allowed the *semantics* (as opposed to simply
validity) of new fields to be described.

### Consider the sync payloads frozen and never change them.

A process where payloads are frozen was rejected because:

* The most naive approach here would not meet the needs of Firefox in the future.

> * A complicated system where we started creating new payload and new collections

Reviewer 4: nitpick (editorial):

```suggestion
* A complicated system where we started creating new payloads and new collections
```

  (ie, freezing "old" schemas but then creating "new" schemas only understood by
  newer clients) could not be conceived in a way that still met the requirements,
  particularly around data-loss for older clients. For example, adding a credit-card
  on a Nightly version but having it be completely unavailable on a release firefox
  isn't acceptable.

### Use separate collections for new data

We could store the new data in a separate collection. For example define a
bookmarks2 collection where each record has the same guid as one in bookmarks alongside any new fields.
Newer clients use both collections to sync.

The pros and cons:

> * Good, because it allows newer clients to sync new data without affecting recent or older clients

Reviewer 4: nitpick (editorial):

```suggestion
* Good, because it allows "new" clients to sync new data without affecting "recent" or "old" clients
```
